// prepack/lib/options.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getRealmOptions = getRealmOptions;
exports.getSerializerOptions = getSerializerOptions;
var defaultOptions = exports.defaultOptions = {}; /**
                                                   * Copyright (c) 2017-present, Facebook, Inc.
                                                   * All rights reserved.
                                                   *
                                                   * This source code is licensed under the BSD-style license found in the
                                                   * LICENSE file in the root directory of this source tree. An additional grant
                                                   * of patent rights can be found in the PATENTS file in the same directory.
                                                   */

function getRealmOptions(_ref) {
  var _ref$compatibility = _ref.compatibility,
      compatibility = _ref$compatibility === undefined ? "browser" : _ref$compatibility,
      mathRandomSeed = _ref.mathRandomSeed,
      _ref$debugNames = _ref.debugNames,
      debugNames = _ref$debugNames === undefined ? false : _ref$debugNames,
      uniqueSuffix = _ref.uniqueSuffix,
      timeout = _ref.timeout,
      strictlyMonotonicDateNow = _ref.strictlyMonotonicDateNow;

  return {
    partial: true,
    compatibility: compatibility,
    mathRandomSeed: mathRandomSeed,
    debugNames: debugNames,
    uniqueSuffix: uniqueSuffix,
    timeout: timeout,
    strictlyMonotonicDateNow: strictlyMonotonicDateNow
  };
}

function getSerializerOptions(_ref2) {
  var _ref2$speculate = _ref2.speculate,
      speculate = _ref2$speculate === undefined ? false : _ref2$speculate,
      _ref2$trace = _ref2.trace,
      trace = _ref2$trace === undefined ? false : _ref2$trace,
      _ref2$singlePass = _ref2.singlePass,
      singlePass = _ref2$singlePass === undefined ? false : _ref2$singlePass,
      _ref2$logStatistics = _ref2.logStatistics,
      logStatistics = _ref2$logStatistics === undefined ? false : _ref2$logStatistics,
      _ref2$logModules = _ref2.logModules,
      logModules = _ref2$logModules === undefined ? false : _ref2$logModules,
      _ref2$delayUnsupporte = _ref2.delayUnsupportedRequires,
      delayUnsupportedRequires = _ref2$delayUnsupporte === undefined ? false : _ref2$delayUnsupporte,
      _ref2$internalDebug = _ref2.internalDebug,
      internalDebug = _ref2$internalDebug === undefined ? false : _ref2$internalDebug;

  return {
    initializeMoreModules: speculate,
    trace: trace,
    singlePass: singlePass,
    logStatistics: logStatistics,
    logModules: logModules,
    delayUnsupportedRequires: delayUnsupportedRequires,
    internalDebug: internalDebug
  };
}
